### PR TITLE
show discount badge in line items in cart

### LINKED ISSUE #1745: show discount on sale items in cart block on front end
If items in the cart are on sale, they should show the sale and full price and discount. This is supported in the cart component but it looks like we need to hook this up in the Store API (or something else is breaking this). Follow up issue to #1741.

Preview cart data in editor shows sale items like this: 
Screen Shot 2020-02-17 at 2 46 23 PM

On front end with real data there's no original/full price or discount badge displayed.

### PR DESCRIPTION
**Pull request by haszari**
Fixes #1745

This PR adds product `prices` data to the cart items API. This is used in the client to calculate any effective discount, by comparing `regular_price * quantity` and `line_total`. Note this changes how discount calculation works in client - previous implementation assumed that `line_subtotal` was exclusive of all discounts (turns out this was incorrect assumption).

~~This PR changes what value is returned by the cart items API for `line_subtotal`. This is documented in the schema as the full undiscounted price, and is used on the client to render the `SAVE $101!` badge.~~

~~I couldn't find this data in the cart_item object provided by core, so I'm manually calculating it using the line item product `regular_price` and quantity. This feels a little like a hack – is there a better way to pull this from the core data structure? Are the situations where this logic will break, or where we need to allow for hooks/filters to run?~~

~~Related to that, the API field `line_subtotal` now has a different meaning to the cart_item field `line_subtotal`. Should we use a different name for the API field to avoid confusion, or do we need to revisit how we represent discounts in the API?~~

Note that there is a discount for the [whole cart in current woo core](https://docs.woocommerce.com/wc-apidocs/class-WC_Legacy_Cart.html#_get_cart_discount_total) - but we need a discount or full price for each line item.

### Screenshots
Screen Shot 2020-02-21 at 3 22 37 PM

### How to test the changes in this Pull Request:

1. Set up some discounted products in your store. There are lots of ways for discounts to apply (e.g. sale price, or coupons), so try a few combinations :)
2. Add some discounted and undiscounted products to your cart.
3. View cart block (add to a page and publish if needed).
4. Confirm all line item details are correct, in particular prices and discount badge.

### CODE REVIEW

**Comment by haszari** [comment]
> there is a problem when you set tax option to `Yes, I will enter prices inclusive of taxes`, now the taxes are considered saving? which is not correct

Good spotting @senadir !  I did a bit of testing to explore this, and we might have other bugs related to tax-inclusive prices. In cart block/API, we always show prices tax-exclusive, which is different to current cart. I've logged an issue for this: #1798

**Comment by haszari** [comment]
Thanks for review @mikejolley @senadir - it didn't feel right modifying what `line_subtotal` means. This PR is now ready for another review 🤞🏻

I've revamped the PR to provide the product `prices` data in the cart endpoint, and then updated the cart js (and the preview cart data used in editor) to calculate discount based on `regular_price` and `quantity`.

Keen to hear your feedback on the api additions in particular, and how they were implemented. Here are some considerations cc @mikejolley:

- Store currency data is in both `prices` and `totals` - should we move this data further up the API response, since it's common to all price info (and in fact, common to all cart items?).  
- I've used an instance of `ProductSchema` in `CartItemSchema` so I can call `get_prices`; there's a bit of logic there which I didn't want to paste-fork. This seems to work ok, but I wondered about making the schema methods `static` since most are pure functions and don't rely on class state (schema classes don't really have state?). Changing schema classes to static is a lot of code impact, but might allow us to more easily reuse schema logic.

Side note: we might want to revisit the API schema docs when this stabilises so it's super clear and easy for woo devs to understand how discounts/coupons/tax are represented in each field. There are a few variations/complexities so it could be tricky getting the docs right.

Note also this doesn't cover every case of coupons X tax - will need to consider this in the open coupon PR (#1790) and look at issues with tax inclusive/exclusive as follow up (#1798).

**Comment by mikejolley** [comment]
> Store currency data is in both prices and totals (e.g. currency_symbol, currency_code) - should we move this data further up the API response, since it's common to all price info (and in fact, common to all cart items?).

For consistency, I like having currency data alongside prices. Whilst all prices in the cart use same currency now, maybe that won't be the case in future, and certainly may not always be the case with other endpoints such as product.

File: src/RestApi/StoreApi/Schemas/CartItemSchema.php [hunk 1]
```diff
@@ -175,7 +189,7 @@ protected function get_properties() {
 					$this->get_store_currency_properties(),
 					[
 						'line_subtotal'     => [
-							'description' => __( 'Line price subtotal (excluding coupons and discounts).', 'woo-gutenberg-products-block' ),
```
Comment (mikejolley) [inline src/RestApi/StoreApi/Schemas/CartItemSchema.php]: Why was this removed? It's still accurate afaik.

File: src/RestApi/StoreApi/Schemas/CartItemSchema.php [hunk 2]
```diff
@@ -224,6 +238,8 @@ public function product_id_exists( $product_id ) {
 	public function get_item_response( $cart_item ) {
 		$product = $cart_item['data'];
 
+		$product_prices = $this->product_schema->get_prices( $product );
```
Comment (mikejolley) [inline src/RestApi/StoreApi/Schemas/CartItemSchema.php]: Doing this is fine 👍 Just a shame we have to duplicate data across endpoints.

**Comment by mikejolley** [comment]
So I think there remains one large point of confusion in this PR and possible in the designs, and that is sale prices vs coupon based discounts. Right now they are being combined which to me doesn't seem correct. cc @nerrad @garymurray 

Showing the sale prices and amount saved on line items I think makes some sense. Essentially the 
`TOTAL` column in this table reflects actual product prices (regular price vs sale price).

Where it gets confusing is when you combine that with coupon based discounts. It makes it appear that products are on sale, when they are not, and sometimes coupon discounts will cause unexpected results here e.g. if a cart coupon gets split across all items in the cart. We're combining product prices with post-discount `line_total`.

IMO if we want to show product prices in this table, we should show regular and sale price only. Discount totals should not change these prices, and instead should only be displayed in the cart totals area. I think that will make the most sense for customers.

So as a rudimentary example, lets say we have a product that is normally $20, but sale price is $10, and then we apply a 50% coupon:

Product Name | Qty 1 | Total <del>$20</del> <ins>$10</ins> (Save 50%).
Subtotal $10
Discounts $5 (50% off coupon)
Total $5

**Comment by garymurray** [comment]
> IMO if we want to show product prices in this table, we should show regular and sale price only. Discount totals should not change these prices, and instead should only be displayed in the cart totals area. I think that will make the most sense for customers.

Yes, agreed.

**Comment by nerrad** [comment]
> Yes, agreed.

Same here.

File: src/RestApi/StoreApi/Schemas/CartItemSchema.php (hunk 1 above)
Comment (haszari) [reply src/RestApi/StoreApi/Schemas/CartItemSchema.php]: True! I've tweaked it (7d50a6f5ebc864cc86a77b1d09502c5026e1423a), as this excludes any coupons that apply to the cart, but does include sale prices. Also clarified the description for `line_total` in 143230765838c525fa19367255926c3fbf6f47c6.

**Comment by haszari** [comment]
> IMO if we want to show product prices in this table, we should show regular and sale price only. Discount totals should not change these prices, and instead should only be displayed in the cart totals area. I think that will make the most sense for customers.

Thanks for clarifying this, it's much clearer to me now. I like the simplicity of this design - coupons apply to carts, and sale prices apply to products (line items). This feels like a nice evolution of the current cart design (which shows the effective price for line items, with coupons applied - see comparison screenshot below). cc @garymurray 

I've updated the PR to implement this approach:

- `Save!` badge is based on sale pricing only.
- `Total` column shows regular price / sale price as appropriate, not effective price (after any cart coupons applied). (Note this is different to existing cart.)
- Renamed the component/class for the sale badge to more strongly associate it with sale prices (vs ambiguous "discount").
- Clarified the API docs for `line_total` and `line_subtotal`.

Please review again @mikejolley @senadir and let me know if I've missed anything :)

Screen Shot 2020-02-25 at 9 58 07 AM
Screen Shot 2020-02-25 at 10 03 09 AM

Note - coupons are not displayed in new cart yet - see #1790 - but they do apply to cart totals API data.

**Comment by mikejolley** [comment]
Thanks for the changes. I'm going to rebase this on master now the coupons PR is in whilst you sleep and test everything. If working, will get this merged :)
